src/ucp/core/ucp_proxy_ep.c

```diff
@@ -183,20 +183,44 @@ static UCS_CLASS_CLEANUP_FUNC(ucp_proxy_ep_t)
     }
 }
 
+static void ucp_proxy_ep_replace_if_owned(uct_ep_h uct_ep, uct_ep_h owned_ep,
+                                          uct_ep_h replacement_ep)
+{
+    ucp_proxy_ep_t *proxy_ep;
+
+    if (uct_ep->iface->ops.ep_destroy == ucp_proxy_ep_destroy) {
+        proxy_ep = ucs_derived_of(uct_ep, ucp_proxy_ep_t);
+        if (proxy_ep->uct_ep == owned_ep) {
+            proxy_ep->uct_ep = replacement_ep;
+        }
+        ucs_assert(replacement_ep != NULL);
+    }
+}
+
 void ucp_proxy_ep_replace(ucp_proxy_ep_t *proxy_ep)
 {
     ucp_ep_h ucp_ep = proxy_ep->ucp_ep;
     ucp_lane_index_t lane;
+    uct_ep_h tl_ep = NULL;
 
     ucs_assert(proxy_ep->uct_ep != NULL);
     for (lane = 0; lane < ucp_ep_num_lanes(ucp_ep); ++lane) {
         if (ucp_ep->uct_eps[lane] == &proxy_ep->super) {
+            ucs_assert(proxy_ep->uct_ep != NULL);    /* make sure there is only one match */
             ucp_ep->uct_eps[lane] = proxy_ep->uct_ep;
+            tl_ep = ucp_ep->uct_eps[lane];
```

Comment:
why do you need to go further? (break is removed)

Reply:
I just thought it should never happen. Ok, let's have an assert

Reply:
right, the assert will make sure of that.